itou/job_applications/models.py

```diff
@@ -1241,7 +1241,7 @@ def __str__(self):
 
     @property
     def pretty_to_state(self):
-        choices = dict(JobApplicationWorkflow.STATE_CHOICES)
+        choices = dict(JobApplicationState.choices)
         return choices[self.to_state]
```

Comment:
```suggestion
        return JobApplicationState(self.to_state).label
```
voir même `self.to_state.label` si on ne s'amuse jamais à passer une `str` au champs `to_state` ?

Reply:
Nope :

```
self = <Variable: 'log.pretty_to_state'>
context = [{'True': True, 'False': False, 'None': None}, {}, {}, {'can_view_personal_information': True, 'can_edit_personal_info...ature', 'sentry_trace_meta': '<meta name="sentry-trace" content="c80b148335d54eb4a3144810486c1aea-9215745855dfdaaa">'}]

    def _resolve_lookup(self, context):
        """
        Perform resolution of a real variable (i.e. not a literal) against the
        given context.

        As indicated by the method's name, this method is an implementation
        detail and shouldn't be called by external code. Use Variable.resolve()
        instead.
        """
        current = context
        try:  # catch-all for silent variable failures
            for bit in self.lookups:
                try:  # dictionary lookup
>                   current = current[bit]
E                   TypeError: 'JobApplicationTransitionLog' object is not subscriptable

.venv/lib/python3.11/site-packages/django/template/base.py:875: TypeError

During handling of the above exception, another exception occurred:

self = <tests.www.apply.test_submit.DirectHireFullProcessTest testMethod=test_hire_as_geiq>

    @pytest.mark.ignore_template_errors
    def test_hire_as_geiq(self):
        """Apply as GEIQ with pre-existing job seeker without previous application"""
        company = CompanyWithMembershipAndJobsFactory(romes=("N1101", "N1105"), kind=CompanyKind.GEIQ)
        job_seeker = JobSeekerFactory()

        user = company.members.first()
        self.client.force_login(user)

        # Step determine the job seeker with a NIR.
        # ----------------------------------------------------------------------

        check_nir_url = reverse("apply:check_nir_for_hire", kwargs={"company_pk": company.pk})
        response = self.client.get(check_nir_url)
        assert response.status_code == 200

        response = self.client.post(check_nir_url, data={"nir": job_seeker.jobseeker_profile.nir, "confirm": 1})
        check_infos_url = reverse(
            "apply:check_job_seeker_info_for_hire", kwargs={"company_pk": company.pk, "job_seeker_pk": job_seeker.pk}
        )
        self.assertRedirects(response, check_infos_url, fetch_redirect_response=False)

        # Step check job seeker infos
        # ----------------------------------------------------------------------

        response = self.client.get(check_infos_url)
        self.assertTemplateNotUsed(response, "approvals/includes/card.html")

        prev_applicaitons_url = reverse(
            "apply:check_prev_applications_for_hire", kwargs={"company_pk": company.pk, "job_seeker_pk": job_seeker.pk}
        )
        self.assertContains(response, prev_applicaitons_url)
        self.assertContains(response, "Éligibilité GEIQ non confirmée")

        # Step check previous applications
        # ----------------------------------------------------------------------

        response = self.client.get(prev_applicaitons_url)
        self.assertTemplateNotUsed(response, "approvals/includes/card.html")
        geiq_eligibility_url = reverse(
            "apply:geiq_eligibility_for_hire", kwargs={"company_pk": company.pk, "job_seeker_pk": job_seeker.pk}
        )
        self.assertRedirects(response, geiq_eligibility_url, fetch_redirect_response=False)

        # Step GEIQ eligibility
        # ----------------------------------------------------------------------
        geiq_criteria_url = reverse(
            "apply:geiq_eligibility_criteria_for_hire",
            kwargs={"company_pk": company.pk, "job_seeker_pk": job_seeker.pk},
        )
        confirmation_url = reverse(
            "apply:hire_confirmation", kwargs={"company_pk": company.pk, "job_seeker_pk": job_seeker.pk}
        )

        response = self.client.get(geiq_eligibility_url)
        self.assertTemplateNotUsed(response, "approvals/includes/card.html")
        response = self.client.post(
            geiq_eligibility_url,
            data={"choice": "True"},
            headers={"hx-request": "true"},
        )
        self.assertRedirects(
            response,
            add_url_params(geiq_criteria_url, {"back_url": check_infos_url, "next_url": confirmation_url}),
            fetch_redirect_response=False,
        )
        htmx_response = self.client.get(geiq_criteria_url, headers={"hx-request": "true"})
        assert htmx_response.status_code == 200

        response = self.client.post(
            add_url_params(geiq_criteria_url, {"back_url": check_infos_url, "next_url": confirmation_url}),
            data={
                "jeune_26_ans": "on",
                "jeune_de_moins_de_26_ans_sans_qualification": "on",
                "proof_of_eligibility": "on",
            },
        )
        self.assertRedirects(response, confirmation_url)

        # Hire confirmation
        # ----------------------------------------------------------------------
        response = self.client.get(confirmation_url)
        self.assertTemplateNotUsed(response, "approvals/includes/card.html")
        self.assertContains(response, "Valider l’embauche")

        hiring_start_at = timezone.localdate()
        post_data = {
            "hiring_start_at": hiring_start_at.strftime(DuetDatePickerWidget.INPUT_DATE_FORMAT),
            "hiring_end_at": "",
            "pole_emploi_id": job_seeker.jobseeker_profile.pole_emploi_id,
            "lack_of_pole_emploi_id_reason": job_seeker.jobseeker_profile.lack_of_pole_emploi_id_reason,
            "answer": "",
            "address_line_1": job_seeker.address_line_1,
            "post_code": job_seeker.post_code,
            "city": self.city.name,
            "city_slug": self.city.slug,
            "prehiring_guidance_days": 3,
            "nb_hours_per_week": 4,
            "planned_training_hours": 5,
            "contract_type": ContractType.APPRENTICESHIP,
            "qualification_type": QualificationType.STATE_DIPLOMA,
            "qualification_level": QualificationLevel.LEVEL_4,
            "hired_job": company.job_description_through.first().pk,
        }
        response = self.client.post(
            confirmation_url,
            data=post_data,
            headers={"hx-request": "true"},
        )
        assert response.status_code == 200
        assert (
            response.headers.get("HX-Trigger") == '{"modalControl": {"id": "js-confirmation-modal", "action": "show"}}'
        )
        post_data = post_data | {"confirmed": "True"}
        response = self.client.post(confirmation_url, headers={"hx-request": "true"}, data=post_data)

        job_application = JobApplication.objects.get(sender=user, to_company=company)
        next_url = reverse("apply:details_for_company", kwargs={"job_application_id": job_application.pk})
        self.assertRedirects(response, next_url, status_code=200, fetch_redirect_response=False)

        assert job_application.job_seeker == job_seeker
        assert job_application.sender_kind == SenderKind.EMPLOYER
        assert job_application.sender_company == company
        assert job_application.sender_prescriber_organization is None
        assert job_application.state == JobApplicationState.ACCEPTED
        assert job_application.message == ""
        assert list(job_application.selected_jobs.all()) == []
        assert job_application.resume_link == ""
        assert job_application.qualification_type == QualificationType.STATE_DIPLOMA
        assert job_application.qualification_level == QualificationLevel.LEVEL_4

        # Get application detail
        # ----------------------------------------------------------------------
>       response = self.client.get(next_url)

tests/www/apply/test_submit.py:2516:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.venv/lib/python3.11/site-packages/django/test/client.py:1049: in get
    response = super().get(path, data=data, secure=secure, headers=headers, **extra)
.venv/lib/python3.11/site-packages/django/test/client.py:465: in get
    return self.generic(
.venv/lib/python3.11/site-packages/django/test/client.py:617: in generic
    return self.request(**r)
tests/utils/test.py:82: in request
    response = super().request(**request)
.venv/lib/python3.11/site-packages/django/test/client.py:1013: in request
    self.check_exception(response)
.venv/lib/python3.11/site-packages/django/test/client.py:743: in check_exception
    raise exc_value
.venv/lib/python3.11/site-packages/django/core/handlers/exception.py:55: in inner
    response = get_response(request)
.venv/lib/python3.11/site-packages/django/core/handlers/base.py:197: in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
.venv/lib/python3.11/site-packages/sentry_sdk/integrations/django/views.py:84: in sentry_wrapped_callback
    return callback(request, *args, **kwargs)
/usr/lib/python3.11/contextlib.py:81: in inner
    return func(*args, **kwds)
.venv/lib/python3.11/site-packages/django/contrib/auth/decorators.py:23: in _wrapper_view
    return view_func(request, *args, **kwargs)
itou/www/apply/views/process_views.py:155: in details_for_company
    return render(request, template_name, context)
.venv/lib/python3.11/site-packages/sentry_sdk/integrations/django/templates.py:102: in render
    return real_render(request, template_name, context, *args, **kwargs)
.venv/lib/python3.11/site-packages/django/shortcuts.py:25: in render
    content = loader.render_to_string(template_name, context, request, using=using)
.venv/lib/python3.11/site-packages/django/template/loader.py:62: in render_to_string
    return template.render(context, request)
.venv/lib/python3.11/site-packages/django/template/backends/django.py:61: in render
    return self.template.render(context)
.venv/lib/python3.11/site-packages/django/template/base.py:171: in render
    return self._render(context)
.venv/lib/python3.11/site-packages/django/test/utils.py:111: in instrumented_test_render
    return self.nodelist.render(context)
.venv/lib/python3.11/site-packages/django/template/base.py:1000: in render
    return SafeString("".join([node.render_annotated(context) for node in self]))
.venv/lib/python3.11/site-packages/django/template/base.py:1000: in <listcomp>
    return SafeString("".join([node.render_annotated(context) for node in self]))
.venv/lib/python3.11/site-packages/django/template/base.py:961: in render_annotated
    return self.render(context)
.venv/lib/python3.11/site-packages/django/template/loader_tags.py:159: in render
    return compiled_parent._render(context)
.venv/lib/python3.11/site-packages/django/test/utils.py:111: in instrumented_test_render
    return self.nodelist.render(context)
.venv/lib/python3.11/site-packages/django/template/base.py:1000: in render
    return SafeString("".join([node.render_annotated(context) for node in self]))
.venv/lib/python3.11/site-packages/django/template/base.py:1000: in <listcomp>
    return SafeString("".join([node.render_annotated(context) for node in self]))
.venv/lib/python3.11/site-packages/django/template/base.py:961: in render_annotated
    return self.render(context)
.venv/lib/python3.11/site-packages/django/template/loader_tags.py:159: in render
    return compiled_parent._render(context)
.venv/lib/python3.11/site-packages/django/test/utils.py:111: in instrumented_test_render
    return self.nodelist.render(context)
.venv/lib/python3.11/site-packages/django/template/base.py:1000: in render
    return SafeString("".join([node.render_annotated(context) for node in self]))
.venv/lib/python3.11/site-packages/django/template/base.py:1000: in <listcomp>
    return SafeString("".join([node.render_annotated(context) for node in self]))
.venv/lib/python3.11/site-packages/django/template/base.py:961: in render_annotated
    return self.render(context)
.venv/lib/python3.11/site-packages/django/template/loader_tags.py:65: in render
    result = block.nodelist.render(context)
.venv/lib/python3.11/site-packages/django/template/base.py:1000: in render
    return SafeString("".join([node.render_annotated(context) for node in self]))
.venv/lib/python3.11/site-packages/django/template/base.py:1000: in <listcomp>
    return SafeString("".join([node.render_annotated(context) for node in self]))
.venv/lib/python3.11/site-packages/django/template/base.py:961: in render_annotated
    return self.render(context)
.venv/lib/python3.11/site-packages/django/template/loader_tags.py:65: in render
    result = block.nodelist.render(context)
.venv/lib/python3.11/site-packages/django/template/base.py:1000: in render
    return SafeString("".join([node.render_annotated(context) for node in self]))
.venv/lib/python3.11/site-packages/django/template/base.py:1000: in <listcomp>
    return SafeString("".join([node.render_annotated(context) for node in self]))
.venv/lib/python3.11/site-packages/django/template/base.py:961: in render_annotated
    return self.render(context)
.venv/lib/python3.11/site-packages/django/template/loader_tags.py:210: in render
    return template.render(context)
.venv/lib/python3.11/site-packages/django/template/base.py:173: in render
    return self._render(context)
.venv/lib/python3.11/site-packages/django/test/utils.py:111: in instrumented_test_render
    return self.nodelist.render(context)
.venv/lib/python3.11/site-packages/django/template/base.py:1000: in render
    return SafeString("".join([node.render_annotated(context) for node in self]))
.venv/lib/python3.11/site-packages/django/template/base.py:1000: in <listcomp>
    return SafeString("".join([node.render_annotated(context) for node in self]))
.venv/lib/python3.11/site-packages/django/template/base.py:961: in render_annotated
    return self.render(context)
.venv/lib/python3.11/site-packages/django/template/defaulttags.py:549: in render
    return self.nodelist.render(context)
.venv/lib/python3.11/site-packages/django/template/base.py:1000: in render
    return SafeString("".join([node.render_annotated(context) for node in self]))
.venv/lib/python3.11/site-packages/django/template/base.py:1000: in <listcomp>
    return SafeString("".join([node.render_annotated(context) for node in self]))
.venv/lib/python3.11/site-packages/django/template/base.py:961: in render_annotated
    return self.render(context)
.venv/lib/python3.11/site-packages/django/template/defaulttags.py:242: in render
    nodelist.append(node.render_annotated(context))
.venv/lib/python3.11/site-packages/django/template/base.py:961: in render_annotated
    return self.render(context)
.venv/lib/python3.11/site-packages/django/template/base.py:1059: in render
    output = self.filter_expression.resolve(context)
<patchy>:4: in resolve
    ???
.venv/lib/python3.11/site-packages/django/template/base.py:842: in resolve
    value = self._resolve_lookup(context)
.venv/lib/python3.11/site-packages/django/template/base.py:885: in _resolve_lookup
    current = getattr(current, bit)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <JobApplicationTransitionLog: 319>

    @property
    def pretty_to_state(self):
>       return self.to_state.label
E       AttributeError: 'str' object has no attribute 'label'
```